KAFKA-5290 docs need clarification on meaning of 'committed' to the log

edoardocomar opened the pull request:
based on conversations with @vahidhashemian @rajinisivaram @apurvam 

The docs didn't make clear that what gets committed and what gets not may depend on the producer acks.

edoardocomar committed:
DOC: clarifications on message committed to the log

based on conversations with @vahidhashemian @rajinisivaram @apurvam

vahidhashemian commented on docs/design.html:
```diff
@@ -298,9 +299,12 @@
-    A message is considered "committed" when all in sync replicas for that partition have applied it to their log. Only committed messages are ever given out to the consumer. This means that the consumer need not worry
-    about potentially seeing a message that could be lost if the leader fails. Producers, on the other hand, have the option of either waiting for the message to be committed or not, depending on their preference for
-    tradeoff between latency and durability. This preference is controlled by the acks setting that the producer uses.
+    We can now more precisely define that a message is considered committed when all in sync replicas for that partition have applied it to their log. 
```
The statement you added to the paragraph above:
> A message is considered committed when all "alive" replicas for that partition have applied it to their log.

And this statement here:
> We can now more precisely define that a message is considered committed when all in sync replicas for that partition have applied it to their log.

seem to imply that "alive" and "in-sync" are the same concepts. But the definition of liveness above defines "alive" as a superset of "in-sync".

edoardocomar replied:
Thanks @vahidhashemian 
I would have liked to use the term `in-sync` in the top paragraph, but it is only introduced later. And in fact the paragraph, refers to the following for details. Perhaps I should have said the same about `committed` ? Please let me know what you think

vahidhashemian commented on docs/design.html:
```diff
@@ -298,9 +299,12 @@
+    Only committed messages are ever given out to the consumer. This means that the consumer need not worry about potentially seeing a message that could be lost if the leader fails. Producers, on the other hand, 
+    have the option of either waiting for the message to be committed or not, depending on their preference for tradeoff between latency and durability. This preference is controlled by the acks setting that the 
+    producer uses.
+    Note that the acks setting does also determine whether a message can be committed or not. Topics have a setting for a "minimum number" of in sync replicas, that it is checked only when the producer requests 
```
>  of in sync replicas, that **it** is checked only when the producer requests 

Perhaps "it" can be removed from this sentence.

vahidhashemian reviewed (commented):
@edoardocomar Thanks for clarifying the paragraph. It looks much better now. I added a couple of minor comments inline.

edoardocomar committed:
KAFKA-5290: docs need clarification on meaning of 'committed' to the log

the producer acks setting is crucial in determining what committed
means.
based on conversations with @vahidhashemian @rajinisivaram @apurvam

edoardocomar commented:
Thanks @vahidhashemian 
please let me know if this version is better

vahidhashemian commented:
Thanks @edoardocomar. It looks good to me. If we want to further clarify what the last added paragraph means we could add an example (similar to the one we discussed). But it's good without the example too.

edoardocomar commented:
Hi @rajinisivaram @ijuma I'd like this change to be merged. 
I did not think it was worth a JIRA, but can open one if needed.

apurvam reviewed (approved):
Thanks for the patch @edoardocomar . The clarification of this surprising behavior looks good to me.

apurvam commented:
I think it may make sense to actually fix this behavior: for instance, we should move the high water mark forward only if the size(ISR) >= min.isr. Otherwise, the high watermark stays the same even if you have producers producing with acks = 1.

rajinisivaram commented:
@hachikuji Can you review this please?

hachikuji commented on docs/design.html:
```diff
@@ -298,9 +298,12 @@
+    Only committed messages are ever given out to the consumer. This means that the consumer need not worry about potentially seeing a message that could be lost if the leader fails. Producers, on the other hand, 
+    have the option of either waiting for the message to be committed or not, depending on their preference for tradeoff between latency and durability. This preference is controlled by the acks setting that the 
+    producer uses.
+    Note that the producer's acks setting also determines whether a message can be committed or not. Topics have a setting for a "minimum number" of in sync replicas, that is checked only when the producer requests 
```
> Note that the producer's acks setting also determines whether a message can be committed or not.

This sentence could be clarified a little. I think the misleading aspect of min.isr that you are trying to capture is the fact that it only guards writes from producers with acks=all. However, once a message is written to the log, whether or not it can become "committed" is independent of producer acks or min.isr.

edoardocomar committed:
KAFKA-5290: docs need clarification on meaning of 'committed' to the log

the producer acks setting is crucial in determining what committed
means.
based on conversations with @vahidhashemian @rajinisivaram @apurvam

edoardocomar commented:
Thanks @hachikuji I've tried to make it clearer. 
Note that I don't think that the behavior of acks is not well documented , rather I found that when a user goes through the docs for the first time and encounters these concepts, a clarification was needed

hachikuji commented on docs/design.html:
```diff
@@ -298,9 +298,13 @@
+    have the option of either waiting for the message to be committed or not, depending on their preference for tradeoff between latency and durability. This preference is controlled by the acks setting that the 
+    producer uses.
+    Note that topics have a setting for the "minimum number" of in-sync replicas, that is checked only when the producer requests the acknowledgment that the message 
+    is written to the full set of in-sync replicas. If a less stringent acknowledgement is requested bt the producer, then the message can be committed, and consumed, 
```
typo: "bt" -> "by"

edoardocomar replied:
oops

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5510/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5495/
Test PASSed (JDK 8 and Scala 2.12).

hachikuji commented on docs/design.html:
```diff
@@ -298,9 +298,13 @@
+    producer uses.
+    Note that topics have a setting for the "minimum number" of in-sync replicas, that is checked only when the producer requests the acknowledgment that the message 
+    is written to the full set of in-sync replicas. If a less stringent acknowledgement is requested bt the producer, then the message can be committed, and consumed, 
+    even if the leader is the only member of the in-sync replica set.
```
Maybe we could make this more general and say "even if the number of in-sync replicas is lower than min.isr"?

edoardocomar replied:
Thanks, @hachikuji I tried to address your comment. Though I think that the case of just the leader remaining up - i.e. no replication - is significant

hachikuji commented on docs/design.html:
```diff
@@ -298,9 +298,13 @@
+    Only committed messages are ever given out to the consumer. This means that the consumer need not worry about potentially seeing a message that could be lost if the leader fails. Producers, on the other hand, 
+    have the option of either waiting for the message to be committed or not, depending on their preference for tradeoff between latency and durability. This preference is controlled by the acks setting that the 
+    producer uses.
+    Note that topics have a setting for the "minimum number" of in-sync replicas, that is checked only when the producer requests the acknowledgment that the message 
```
Can we drop the word "only"? It makes the check sound weak in some way, but I think it's a matter of perspective. Kafka puts the burden on producers which require higher durability to ensure their messages reach the topic.

edoardocomar replied:
I am not sure, that 'only' was there in the original docs and I always took it as meaning that messages written to the local log but not yet considered 'committed' (i.e. not replicated to the isr) could not be consumed yet.

hachikuji replied:
Sorry, the "only" I'm referring to is the one on this line: 
> Note that topics have a setting for the "minimum number" of in-sync replicas that is checked *only* when the producer requests ...

edoardocomar replied:
Agree, that's redundant thanks

hachikuji reviewed (commented):
Thanks for the patch. Left a couple additional comments.

edoardocomar committed:
KAFKA-5290: docs need clarification on meaning of 'committed' to the log

the producer acks setting is crucial in determining what committed
means.
based on conversations with @vahidhashemian @rajinisivaram @apurvam

edoardocomar committed:
KAFKA-5290: docs need clarification on meaning of 'committed' to the log

the producer acks setting is crucial in determining what committed
means.
based on conversations with @vahidhashemian @rajinisivaram @apurvam

hachikuji reviewed (approved):
Thanks for the patch. LGTM.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5523/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5508/
Test PASSed (JDK 8 and Scala 2.12).
